{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"L03_2_python_library_biopython.HW.ANS.ipynb","provenance":[],"collapsed_sections":["P4mLXvaysCMv","r1-JPjY_j2-u","h4jxlQQfLNYb","rIuNuWpXpV3g","va8dgXEE6hRG"]},"kernelspec":{"name":"python3","display_name":"Python 3"}},"cells":[{"cell_type":"markdown","metadata":{"id":"gUyn5DvT5SdH"},"source":["<img src=\"https://lh3.googleusercontent.com/pw/AM-JKLVhTn_UySwMdfMwXvoq8l3VN7IkrY9cwtH2YJVMxAlMznUBWC9IpFtgPRIyfAXru4oykkYD-1WjWi0Ao5XgkB9JICvzDBcfn0L_5X2_KOOppsURK5DfSifCC-s7Vx5oQrBUn_BNWn_hfAPdhlVbKQGE=w1097-h235-no?authuser=0\" alt=\"2021年度ゲノム情報解析入門\" height=\"100px\" align=\"middle\">\n","\n","<div align=\"right\"><a href=\"https://github.com/CropEvol/lecture#section2\">実習表ページに戻る</a></div>"]},{"cell_type":"markdown","metadata":{"id":"V84X2UUCIFWb"},"source":["# Pythonライブラリを使う - Biopython - [課題]（11月2日版）\n","\n","実習テキストは [こちら](https://colab.research.google.com/github/CropEvol/lecture/blob/master/textbook_2021/L03_python_library_biopython.ipynb)"]},{"cell_type":"code","metadata":{"id":"bz-qgNd1eDXl"},"source":["### ＊重要＊\n","### 最初にこのセルを実行してください。\n","### サンプルデータをダウンロードやライブラリのインストールをおこないます。\n","\n","### サンプルデータのダウンロード\n","!wget -O \"mini_genome.fasta\" https://raw.githubusercontent.com/CropEvol/lecture/master/data/mini_genome.fasta\n","\n","### Biopythonのインストール\n","!pip install biopython"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"VGMdxfrFrCTD"},"source":["## 課題1\n","\n","　サンプルデータ[`mini_genome.fasta`](https://raw.githubusercontent.com/CropEvol/lecture/master/data/mini_genome.fasta)は、仮想生物のゲノム配列です（9本の染色体chromosome_01〜chromosome09の配列）。\n","\n","　この仮想生物のゲノムサイズ（各染色体の長さの合計値）を調べてください。"]},{"cell_type":"code","metadata":{"id":"KXpSAIxbrdIP"},"source":["### BiopythonのSeqIOを準備\n","from Bio import SeqIO\n","\n","### ファイルを開く\n","f = open(\"mini_genome.fasta\")\n","\n","### FASTA形式で読み込み\n","recs = SeqIO.parse(f, \"fasta\")\n","\n","### 1データ（1染色体）ずつ取り出す\n","for rec in recs:\n","  print(rec.id)\n","  print(rec.seq)\n","\n","### ファイルを閉じる\n","f.close()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"P4mLXvaysCMv"},"source":["### 解答例: `sum`関数を使わない方法"]},{"cell_type":"markdown","metadata":{"id":"n070vc3snHHA"},"source":["重要なポイント: \n","- ゲノムサイズ用の変数を事前に準備する\n","  - 11行目: `chrom_size = 0`\n","- 染色体の長さを取得した際に、上記の変数にその値を追加する（変数の値を更新する）\n","  - 19行目: `chrom_size = chrom_size + len(rec.seq)`"]},{"cell_type":"code","metadata":{"id":"FsE43IierENp"},"source":["### BiopythonのSeqIOを準備\n","from Bio import SeqIO\n","\n","### ファイルを開く\n","f = open(\"mini_genome.fasta\")\n","\n","### FASTA形式で読み込み\n","recs = SeqIO.parse(f, \"fasta\")\n","\n","### ゲノムサイズ用の変数（初期値0）\n","chrom_size = 0\n","\n","### 1データ（1染色体）ずつ取り出す\n","for rec in recs:\n","  #print(rec.id)\n","  #print(rec.seq)\n","\n","  ### 調べた染色体の長さをゲノムサイズ用変数に追加\n","  chrom_size = chrom_size + len(rec.seq)\n","\n","### 結果を表示\n","print(chrom_size)\n","\n","### ファイルを閉じる\n","f.close()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"r1-JPjY_j2-u"},"source":["### 解答例: `sum`関数を使う方法"]},{"cell_type":"markdown","metadata":{"id":"pKbVVO_hk5Rp"},"source":["重要なポイント: \n","- 染色体の長さを一時保管するためのリストを事前に準備する\n","  - 11行目: `chrom_size_list = []`\n","- 染色体の長さを取得した際に、一時保管用リストにその値を追加する\n","  - 19行目: `chrom_size_list.append(len(rec.seq))`\n","- `sum`関数でリストの合計値を得る\n","  - 22行目: `print(sum(chrom_size_list))`\n","\n","\n","```python\n","# リストに値を追加する方法1\n","# （append関数でリストへの値の追加をおこなえる）\n","リスト.append(値)\n","\n","# 例1\n","fruits = [\"apple\", \"orange\"]\n","fruits.append(\"grape\")\n","print(fruits) # 出力結果=> [\"apple\", \"orange\", \"grape\"]\n","```\n","\n","```python\n","# リストに値を追加する方法2\n","# （リストの足し算で、リストの結合をおこなえる）\n","リスト.append(値)\n","\n","# 例2\n","fruits = [\"apple\", \"orange\"]\n","fruits = fruits + [\"grape\"]\n","print(fruits) # 出力結果=> [\"apple\", \"orange\", \"grape\"]\n","```"]},{"cell_type":"code","metadata":{"id":"JmypBP6Pj4WE"},"source":["### BiopythonのSeqIOを準備\n","from Bio import SeqIO\n","\n","### ファイルを開く\n","f = open(\"mini_genome.fasta\")\n","\n","### FASTA形式で読み込み\n","recs = SeqIO.parse(f, \"fasta\")\n","\n","### 各染色体の長さを一時保管するためのリスト変数（初期値0）\n","chrom_size_list = []\n","\n","### 1データ（1染色体）ずつ取り出す\n","for rec in recs:\n","  #print(rec.id)\n","  #print(rec.seq)\n","\n","  ### 調べた染色体の長さをリストに追加\n","  chrom_size_list.append(len(rec.seq))\n","\n","  ### 調べた染色体の長さをリストに追加（以下でもOK）\n","  #chrom_size_list = chrom_size_list + [len(rec.seq)]\n","\n","### リスト内の値の合計値を調べる\n","print(sum(chrom_size_list))\n","\n","### ファイルを閉じる\n","f.close()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"aTZcsDQYKY5y"},"source":["## 課題2\n","\n","　各染色体に、塩基Gがそれぞれ何個含まれているか調べてください。\n","\n","``` python\n","# 文字列中に含まれる任意文字の個数の調べ方\n","文字列.count(\"任意文字\")\n","\n","# 例: 文字列\"apppppple\"中に含まれる文字\"p\"の個数\n","\"apppppple\".count(\"p\")\n","```"]},{"cell_type":"code","metadata":{"id":"fD3gwSd1Ko-O"},"source":["### BiopythonのSeqIOを準備\n","from Bio import SeqIO\n","\n","### ファイルを開く\n","f = open(\"mini_genome.fasta\")\n","\n","### FASTA形式で読み込み\n","recs = SeqIO.parse(f, \"fasta\")\n","\n","### 1データ（1染色体）ずつ取り出す\n","for rec in recs:\n","  print(rec.id)\n","  print(rec.seq)\n","\n","### ファイルを閉じる\n","f.close()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"h4jxlQQfLNYb"},"source":["### 解答例"]},{"cell_type":"code","metadata":{"id":"GNMKPlkTLNHQ"},"source":["### BiopythonのSeqIOを準備\n","from Bio import SeqIO\n","\n","### ファイルを開く\n","f = open(\"mini_genome.fasta\")\n","\n","### FASTA形式で読み込み\n","recs = SeqIO.parse(f, \"fasta\")\n","\n","### 1データ（1染色体）ずつ取り出す\n","for rec in recs:\n","  #print(rec.id)\n","  #print(rec.seq)\n","\n","  ### 塩基Gの個数\n","  count_G = rec.seq.count(\"G\")\n","  print(rec.id, count_G)\n","\n","### ファイルを閉じる\n","f.close()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"FwAUDZJESj3M"},"source":["## 課題3\n","\n","　仮想生物の各染色体のGC含量（塩基GとCの割合）を調べてください。\n","\n","```\n","# ひとつの染色体中のGC含量\n","GC含量 = (塩基Gの個数 + 塩基Cの個数) / 染色体の長さ\n","```\n","\n"]},{"cell_type":"code","metadata":{"id":"IyZwP-qNeeeV"},"source":["### BiopythonのSeqIOを準備\n","from Bio import SeqIO\n","\n","### ファイルを開く\n","f = open(\"mini_genome.fasta\")\n","\n","### FASTA形式で読み込み\n","recs = SeqIO.parse(f, \"fasta\")\n","\n","### 1データ（1染色体）ずつ取り出す\n","for rec in recs:\n","  print(rec.id)\n","  print(rec.seq)\n","\n","### ファイルを閉じる\n","f.close()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"rIuNuWpXpV3g"},"source":["### 解答例"]},{"cell_type":"code","metadata":{"id":"qwdzlotNpXwZ"},"source":["### BiopythonのSeqIOを準備\n","from Bio import SeqIO\n","\n","### ファイルを開く\n","f = open(\"mini_genome.fasta\")\n","\n","### FASTA形式で読み込み\n","recs = SeqIO.parse(f, \"fasta\")\n","\n","### 1データ（1染色体）ずつ取り出す\n","for rec in recs:\n","  #print(rec.id)\n","  #print(rec.seq)\n","  \n","  ### Gの個数\n","  count_G = rec.seq.count(\"G\")\n","  ### Cの個数\n","  count_C = rec.seq.count(\"C\")\n","  ### 染色体の長さ\n","  chrom_size = len(rec.seq)\n","  ### GC含量\n","  GC_content = (count_G + count_C) / chrom_size\n","\n","  ### 結果を表示\n","  print(rec.id, GC_content)\n","\n","### ファイルを閉じる\n","f.close()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"7ISQlWBZlRcC"},"source":["## 課題4\n","<small>※ この課題は回答不要です。</small>\n","\n","次のワードを調べてください。\n","- Single Nucleotide Polymorphism（SNP; 一塩基多型）  \n","　2つ以上の塩基配列を比較したときに見つかる一塩基の違いのことを**SNP**と言います。  \n","　次世代シーケンス解析では、リファレンス（参照配列）にリードをアライメントし、リファレンスとリードの間の違い（下図では、個体Aと個体Bの間の違い）を探し、SNPを同定します。見つかったSNP情報をもとにゲノムワイドな解析をおこなうことができます。\n","\n","<img src=\"https://lh3.googleusercontent.com/pw/AM-JKLViNK-9cLg8wOMokZH2-AWh2Mrlsm9BOT4oGU4C9QFFrU___O30S7jiUUavzx5mA_TyQnxrCNpn3EpFGSFtEt1mA311teSyWV4QgUAmOaUIiW8xLk6KV9zqoYOKgjK1x_qS1tD1WlYKGOFP9pXBH1H4=w2154-h536-no?authuser=0\" alt=\"NGS_processing\" height=\"250px\">\n","\n","- Read depth（リードデプス）  \n","　ある塩基座にアライメントされているリードの本数のことを**リードデプス**と言います。  \n","　上図の「多型検出」にある赤枠部分の塩基座を例にすると、その塩基座には8本のリードがアライメントされています。したがって、リードデプスは8です。"]},{"cell_type":"markdown","metadata":{"id":"va8dgXEE6hRG"},"source":["---\n","## 課題提出について\n","\n","**提出期限: 11月8日（月）**\n","\n","　次の手順で、課題を提出してください。\n","\n","1. 課題取り組み後のノートブックの共有リンクを発行する。  \n","  共有リンク発行手順:\n","  1. ノートブックを保存する\n","    - 「ファイル」 > 「保存」または「ドライブにコピーを保存」\n","  1. 共有リンクを発行する  \n","    - 右上の「共有」ボタンをクリックする\n","    - 「リンクを知っている全員に変更」をクリックする\n","    - 権限が「閲覧者」になっているか確認する\n","    - 共有リンクをコピーする\n","1. [PandAのゲノム情報解析入門ページ](https://panda.ecs.kyoto-u.ac.jp/portal/site/2020-111-5323-000)にアクセスする\n","1. 今回の課題提出ページに移動する  \n","  - 「課題」 > 「課題（11月2日）」\n","1. 上で発行した共有リンクを提出フォームに貼り付ける\n","1. 提出する\n","\n","<div align=\"right\"><a href=\"https://github.com/CropEvol/lecture#section2\">実習表ページに戻る</a></div>\n"]}]}