{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"L04_python_genomics.ipynb","provenance":[],"collapsed_sections":["dzuyb6om2Ylz","RCeLpCFa2-_Q","yo5BPd9u8uSQ"],"authorship_tag":"ABX9TyNHRFkIv7RL34UKOEixMty1"},"kernelspec":{"name":"python3","display_name":"Python 3"}},"cells":[{"cell_type":"markdown","metadata":{"id":"rEYhXKVThJr6","colab_type":"text"},"source":["<img src=\"https://github.com/CropEvol/lecture/blob/master/ILAS_2020/01/images/banner.png?raw=true\" alt=\"ILASセミナー「ゲノム博物学入門」\" height=\"100px\" align=\"middle\">"]},{"cell_type":"markdown","metadata":{"id":"JVbvXrd4hMLF","colab_type":"text"},"source":["# データ解析入門 - ゲノム解析入門 -\n"]},{"cell_type":"markdown","metadata":{"id":"ikCYmUYdyj3B","colab_type":"text"},"source":["## Introduction\n","\n","　今回は、ゲノムデータ解析をおこないます。**MutMap** (Abe et al., *Nat. Biotechnol.*, 2012) と呼ばれる解析手法を使って、イネの葉が「淡緑色（pale green）」となる原因遺伝子を探します。\n","\n","<img src=\"https://github.com/CropEvol/lecture/blob/master/textbook_2019/images/mutmap04.jpg?raw=true\" alt=\"mutmap04\" height=\"230px\">\n","\n","　MutMap は、突然変異形質をもつ複数の個体のDNAをひとつにまとめてシーケンシングし、そのシーケンスのデータ解析をおこない、形質に関わる遺伝子がゲノム上のどこにあるかを探索する解析手法（遺伝子マッピング手法）の一つです。\n","\n","次のような手順で解析を行います。\n","\n","1. ある系統（オリジナル系統）から突然変異系統を作出します。\n","1. 突然変異系統とオリジナル系統を交配します。その交配後第二世代（F2集団）では、オリジナル系統と同じ形質を持った個体と突然変異形質を持った個体が分離してきます。\n","1. 突然変異形質をもった個体のDNAを集めます（以下バルクDNA）。\n","1. オリジナル系統のDNAとバルクDNAをシーケンスします。\n","1. 両者のシーケンスを比較すると、突然変異の位置と、オリジナル系統の塩基や突然変異の塩基の種類がわかります。また、それぞれの塩基がいくつあるか（オリジナル系統と同じ塩基の個数と突然変異系統と同じ塩基の個数）といった情報も得られます。\n","1. 突然変異形質を持つF2個体は、形質に関わる何らかの変異を共通で持っているはずです。そのため、形質に関わる変異箇所では、突然変異系統の塩基のみ検出されることが期待できます。そこで、それぞれの一塩基多型（**SNP**）で、突然変異塩基の割合（**SNP-index**と呼ばれるもの）を調べます。\n","1. ゲノム全体にわたってSNP-index調べて、形質の原因となる遺伝子領域を探します。\n","\n","<img src=\"https://github.com/CropEvol/lecture/blob/master/textbook_2019/images/mutmap01.jpg?raw=true\" alt=\"mutmap01\" height=\"400px\">\n","\n","<img src=\"https://github.com/CropEvol/lecture/blob/master/textbook_2019/images/mutmap02.png?raw=true\" alt=\"mutmap02\" height=\"200px\">\n","\n","　MutMapについてより詳しく知りたい方は、下記の論文をご覧ください。\n","\n","> Abe, A., Kosugi, S., Yoshida, K., Natsume, S., Takagi, H., Kanzaki, H., Matsumura, H., Yoshida, K., Mitsuoka, C., Tamiru, M., Innan, H., Cano, L., Kamoun, S., Terauchi, R. (2012). [Genome sequencing reveals agronomically important loci in rice using MutMap.](https://www.nature.com/articles/nbt.2095)  Nature biotechnology, 30(2), 174.\n","\n","　このテキストでは、下記のような、イネ第10染色体のデータのみを扱います。\n","\n","<img src=\"https://github.com/CropEvol/lecture/blob/master/textbook_2019/images/mutmap03.jpg?raw=true\" alt=\"mutmap03\" height=\"350px\">"]},{"cell_type":"markdown","metadata":{"id":"jag100dA0YJN","colab_type":"text"},"source":["　実習を始める前に、サンプルデータのダウンロードと、Python (pandas)での読み込みをおこなってください。"]},{"cell_type":"code","metadata":{"id":"UgWE71fB01Rp","colab_type":"code","colab":{}},"source":["\"\"\"\n","＊重要＊\n","最初にこのセルを実行してください。サンプルファイルをダウンロードします。\n","\"\"\"\n","!wget -O mutmap_bulk.txt https://raw.githubusercontent.com/CropEvol/lecture/master/data/mutmap_bulk.txt"],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"wV0xbwbp0utz","colab_type":"code","colab":{}},"source":["import pandas as pd   # pandasを準備する \n","\n","# ファイルを読み込む\n","# header=None（列名行なし）により、列名として通し番号が付けられます。\n","# names=[リスト]で、通し番号の代わりに列名を付けられます。\n","df = pd.read_csv('mutmap_bulk.txt', sep=\"\\t\", \n","                 header=None, names=['chr', 'pos', 'ref_nucl', 'alt_nucl', 'ref_N', 'alt_N'])\n","\n","# 表示\n","df"],"execution_count":0,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"1pjSCuJjCy6k","colab_type":"text"},"source":["## 今回の実習内容\n","\n","1. SNP-indexを求める\n","1. 条件にあうデータを抽出する\n","1. SNP-indexのグラフを描く\n","1. Sliding window解析（発展）\n"]},{"cell_type":"markdown","metadata":{"id":"0mGZ-8yq0XZE","colab_type":"text"},"source":["##1. SNP-indexを求める\n","\n","　SNP-indexを算出してみましょう。計算に使う列を取り出し、次の計算をおこなうだけです。\n","```\n","# SNP-index計算式\n","SNP-index = alt_N / (ref_N + alt_N)\n","```\n","\n"]},{"cell_type":"code","metadata":{"id":"i4AOFEq83LBT","colab_type":"code","colab":{}},"source":["# 分母: ref_N + alt_N\n","df[\"ref_N\"] + df[\"alt_N\"]"],"execution_count":0,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Mpe5xj-q2LzQ","colab_type":"text"},"source":["### 実習1\n","\n","　SNP-indexの計算式を記述して、データフレームに新しい列 `snp_index` を追加してください。"]},{"cell_type":"code","metadata":{"id":"BrDTlRlgykTk","colab_type":"code","colab":{}},"source":["df['snp_index'] = \n","\n","df"],"execution_count":0,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"dzuyb6om2Ylz","colab_type":"text"},"source":["#### 解答例"]},{"cell_type":"code","metadata":{"id":"29C3HVc52ZQ_","colab_type":"code","colab":{}},"source":["df['snp_index'] = df['alt_N'] / (df['ref_N'] +df['alt_N'])\n","\n","# 表示\n","df"],"execution_count":0,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"nVQ8OlZr2k95","colab_type":"text"},"source":["## 2. 条件に合うデータを抽出する\n","\n","　前回のテキストで、条件に合うデータの抽出方法を学習しました。おさらいですが、コードの書き方は以下の通りです。\n","\n","```python\n","データフレーム[ (条件式)　]   # 条件式が一つの場合、丸括弧()は省略可能です\n","\n","データフレーム[ (条件式1) & (条件式2)　]   # 条件2つ（AND条件）\n","データフレーム[ (条件式1) | (条件式2)　]   # 条件2つ（OR条件）\n","```\n","　確認として例を見てみましょう。次のセルを実行してください。\n"]},{"cell_type":"code","metadata":{"id":"Edac50Xi2z6N","colab_type":"code","colab":{}},"source":["df['ref_nucl']=='G'         # True/Falseリストが得られる\n","#df[ df['ref_nucl']=='G' ]  # Trueのデータ行が得られる"],"execution_count":0,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"-oiWJvJr20eB","colab_type":"text"},"source":["### 実習2\n","\n","　SNP-indexが0.8以上のデータ行を取り出してください。"]},{"cell_type":"code","metadata":{"id":"_sUjzf7R2_Tb","colab_type":"code","colab":{}},"source":[""],"execution_count":0,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"RCeLpCFa2-_Q","colab_type":"text"},"source":["#### 解答例"]},{"cell_type":"code","metadata":{"id":"Cbg34TZW26YV","colab_type":"code","colab":{}},"source":["df[df['snp_index'] >= 0.8]"],"execution_count":0,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"mGl6312D4oap","colab_type":"text"},"source":["## 3. SNP-indexのグラフを描く\n","\n","　Pythonのグラフ描画で有名なライブラリ **Matplotlib** を使ってグラフを描いてみましょう。Matplotlibの基本は、前回のテキストで勉強したとおり、「透明なシート（レイヤー）を重ねて、ひとつのグラフを作っていく」です。\n","\n","<img src=\"https://github.com/CropEvol/lecture/blob/master/textbook_2019/images/drawing_graph.png?raw=true\" alt=\"basis_of_matplotlib_drawing\" height=\"200px\">\n","\n","　ここでは、x軸に染色体上の位置、y軸にSNP-indexをとったグラフ（散布図）を作成してみましょう。\n"]},{"cell_type":"code","metadata":{"id":"Tq-mTOzY4wa0","colab_type":"code","colab":{}},"source":["import matplotlib.pyplot as plt   # ライブラリを読み込む\n","\n","df['snp_index'] = df['alt_N'] / (df['ref_N'] + df['alt_N'])  \n","\n","x = df['pos']         # x軸データ\n","y = df['snp_index']   # y軸データ\n","\n","fig = plt.figure(figsize=[12,4])                         # グラフの大きさを設定（省略可）\n","#plt.scatter(x, y, color='gray')                          # 散布図\n","#plt.title('SNP-index on chromosome 10', fontsize=24)  # グラフタイトル\n","#plt.xlabel('Position (x 10 Mb)', fontsize=16)  # x軸ラベル\n","#plt.ylabel('SNP-index', fontsize=16)           # y軸ラベル\n","plt.show()                                    # グラフ表示（省略可）"],"execution_count":0,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"zU-Rdi6350Lj","colab_type":"text"},"source":["### 実習3\n","\n","　全SNP-indexの散布図のうち、SNP-indexが0.8以上の点を赤色にしてください。\n"]},{"cell_type":"code","metadata":{"id":"BZVrZ_OO5ESk","colab_type":"code","colab":{}},"source":["import matplotlib.pyplot as plt  \n","\n","df['snp_index'] = df['alt_N'] / (df['ref_N'] + df['alt_N'])  \n","\n","x = df['pos']        \n","y = df['snp_index']   \n","\n","fig = plt.figure(figsize=[12,4])\n","\n","# 全SNP-indexを表示\n","plt.scatter(x, y, color='gray') \n","plt.title('SNP-index on chromosome 10', fontsize=24)\n","plt.xlabel('Position (x 10 Mb)', fontsize=16)\n","plt.ylabel('SNP-index', fontsize=16)           \n","\n","# SNP-indexが0.8以上大きいものを赤色で表示\n","# （上で作成した散布図に、赤色の散布図を重ねる）\n","\n","# 元のデータフレームdfを上書きしないように、別名のデータフレームを作成\n","df1 = df[ df['snp_index'] >= 0.8 ]   \n","\n","x1 = df1[]   # x軸データを[]内に書き足す\n","y1 = df1[]   # y軸データを[]内に書き足す\n","plt.scatter()   # 赤色で表示させるためのコードを()内に書き足す\n","\n","plt.show()      # グラフ表示"],"execution_count":0,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"yo5BPd9u8uSQ","colab_type":"text"},"source":["#### 解答例"]},{"cell_type":"code","metadata":{"id":"PVrL41sF8wOG","colab_type":"code","colab":{}},"source":["import matplotlib.pyplot as plt  \n","\n","df['snp_index'] = df['alt_N'] / (df['ref_N'] + df['alt_N'])  \n","\n","x = df['pos']        \n","y = df['snp_index']   \n","\n","fig = plt.figure(figsize=[12,4])    \n","plt.scatter(x, y, color='gray') \n","plt.title('SNP-index on chromosome 10', fontsize=24)\n","plt.xlabel('Position (x 10 Mb)', fontsize=16)\n","plt.ylabel('SNP-index', fontsize=16)           \n","\n","# SNP-indexが0.8より大きいものを表示\n","# \"df\"だとオリジナルのデータフレームが上書きされてしまうため、別名のデータフレームを作成\n","df1 = df[ df['snp_index'] >= 0.8 ]   \n","\n","x1 = df1[\"pos\"]   # x軸データを[]内に書き足す\n","y1 = df1[\"snp_index\"]   # y軸データを[]内に書き足す\n","plt.scatter(x1, y1, color='red')   # 赤色で表示させるためのコードを()内に書き足す\n","\n","plt.show()      # グラフ表示"],"execution_count":0,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"iQzWxtIy6ehh","colab_type":"text"},"source":["## 4. Sliding Window解析（発展）\n","\n","　MutMap解析では、SNP-indexの区間平均をゲノムに全体にわたって調べる解析（**Sliding Window解析**）というのをおこない、形質にかかわるゲノム領域を特定します。\n","\n","　形質にかかわる領域付近では、直接の原因となる塩基座位だけでなく、周囲の塩基座位のSNP-indexも1に近づきます。\n","\n","<img src=\"https://github.com/CropEvol/lecture/blob/master/textbook_2019/images/sliding_window_ja.jpg?raw=true\" alt=\"sliding_window\" height=\"400px\">\n","\n","　ここでは上の図のように、Chromosome 10上のSNP-indexの推移を調べてみましょう。Sliding window解析をおこなうプログラムはすでに書かれており、実行するだけでOKです。"]},{"cell_type":"markdown","metadata":{"id":"4rU8FWY46h40","colab_type":"text"},"source":["### 解析方法\n","\n","　SNP-indexのSliding Window解析は次のような手順で解析します。\n","\n","    1.区間サイズとステップサイズを決める。\n","    2.各区間の代表値（ここでは区間の中央値）用のリストと、区間内の平均SNP-index用のリストを用意する。\n","    3.while構文（for構文とは別のループ処理をおこなう構文）を使って全ての区間を調べていく。\n","        区間内に含まれるデータを抽出する。\n","        抽出したデータからSNP-indexの平均値を算出する。\n","        区間の中央値と算出した平均SNP-indexを、それぞれのリストに追加する。\n","        ゲノムの全長を全て調べたらwhileから出る\n","    4.x軸に中央値、y軸に平均SNP-indexの値をとったグラフ（折れ線グラフなど）を作成する。\n"]},{"cell_type":"code","metadata":{"id":"gqUokazq6j2l","colab_type":"code","colab":{}},"source":["###### Sliding Window解析 ######\n","#---  NumPyライブラリの読み込み ---\n","import numpy as np\n","\n","#--- 区間サイズとステップサイズ ---\n","CHROM_SIZE = 23207287       # Chromosome 10　の全長 (bp)\n","WIN_SIZE       = 1 * 1000 * 1000     #  1 Mb = 1000 kb = 1,000,000 bp\n","STEP_SIZE     = 0.2 * 1000* 1000     #  0.2 Mb = 200 kb = 200,000 bp\n","\n","#--- 区間データ用のリストを準備 ---\n","win_position  = []  # 区間中央値用リスト\n","win_snpindex = []  # 平均SNP-index用リスト\n","\n","#--- 全ての区間を調べる ---\n","\"\"\"\n","/// 調べる区間 ///\n","start, end\n","0, 0+1000 (kb)\n","200, 200+1000\n","400, 400+1000\n","  :\n","  :\n","\n","/// WIN_SIZEとSTEP_SIZEを使って表現すると... ///\n","繰り返し数:　n = 0, 1, 2, ...\n","\n","start = STEP_SIZE * n  \n","end = start + WIN_SIZE\n","\n","\n","end > CHROM_SIZEになったとき、繰り返しを終える。\n","\n","このように、繰り返し数を指定しない場合は、for文ではなく、while文を使用します。\n","\"\"\"\n","\n","n = 0 # 繰り返し数\n","while True:\n","    \n","    #--- 区間のstartとend position ---\n","    start = STEP_SIZE * n \n","    end   = start + WIN_SIZE\n","    \n","    #--- 区間の中央値をリストに追加する ---\n","    p = (start + end) / 2\n","    win_position.append(p)\n","    \n","    #--- 区間内データを抽出 ---\n","    sub = df[(df['pos'] >= start) & (df['pos'] < end)]\n","    \n","    #--- SNP-indexの平均値を算出 ---\n","    i = sub['snp_index'].mean()\n","    win_snpindex.append(i)\n","        \n","    #--- 繰り返し数を+1 ---\n","    n += 1\n","    \n","    #--- 全ての区間を調べた時、whileから出る ---\n","    if end > CHROM_SIZE:\n","        break\n","\n","#--- グラフ作成  ---\n","fig = plt.figure(figsize=[12,4])\n","plt.scatter(x, y, color='gray')      # 全データ\n","plt.title('SNP-index on chromosome 10', fontsize=24)  # グラフタイトル\n","plt.xlabel('Position (x 10 Mb)', fontsize=16)  # x軸ラベル\n","plt.ylabel('SNP-index', fontsize=16)        # y軸ラベル\n","\n","#--- 平均SNP-indexの折れ線グラフ（重ね描き） ---\n","plt.plot(win_position, win_snpindex, color='blue')"],"execution_count":0,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Vh9BPKyb6o1W","colab_type":"text"},"source":["　平均SNP-indexが高いゲノム領域（22,482,910bpの辺り）には、「Chlorophyllide a oxygenase (Os10t0567400)」という遺伝子が存在しています。この遺伝子のノックアウト形質転換体（遺伝子の機能を壊した個体）は、葉の色が淡緑色になることが明らかになっています([Abe et al., 2012](https://www.nature.com/articles/nbt.2095))。"]},{"cell_type":"markdown","metadata":{"id":"zlXIiDkG6ppd","colab_type":"text"},"source":["---\n","\n","## まとめ\n","\n","　ゲノム解析の一例として、**MutMap解析**をおこないました。**pandas**を使って、列同士の数値計算（**SNP-index**の計算）をおこないました。また、**Matplotlib**で、データをグラフ化しました。やや複雑なプログラムにはなりますが、ゲノム全体に渡って区間平均を求めるようなゲノム解析（**Sliding window解析**）をおこなえることを学びました。\n","\n","　今回扱ったデータは、272行の小規模なデータ（イネ第10染色体のみのデータ）でしたが、より大規模なデータ（数万行、数10万行のデータ）も同様に解析可能です。\n","\n"]},{"cell_type":"markdown","metadata":{"id":"g5I38_A960zD","colab_type":"text"},"source":["---\n","\n","## 練習問題\n","　第12染色体のデータを使って、SNP-indexの計算、グラフ描画をおこなってください。"]},{"cell_type":"code","metadata":{"id":"XfrscRLk7tJV","colab_type":"code","colab":{}},"source":["\"\"\"\n","＊重要＊\n","このセルを実行してください。練習問題用のサンプルファイルをダウンロードします。\n","\"\"\"\n","!wget -O mutmap_bulk_chr12.txt https://raw.githubusercontent.com/CropEvol/lecture/master/data/mutmap_for_hw.txt"],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"zlOk8xpe7-3l","colab_type":"code","colab":{}},"source":["import pandas as pd   # pandasを準備する \n","\n","# ファイルを読み込む\n","# header=None（列名行なし）により、列名として通し番号が付けられます。\n","# names=[リスト]で、通し番号の代わりに列名を付けられます。\n","df = pd.read_csv('mutmap_bulk_chr12.txt', sep=\"\\t\", \n","                 header=None, names=['chr', 'pos', 'ref_nucl', 'alt_nucl', 'ref_N', 'alt_N'])\n","\n","# 表示\n","df"],"execution_count":0,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"MeGcf-Iz8Ker","colab_type":"text"},"source":["### 練習問題1. SNP-indexを求める"]},{"cell_type":"code","metadata":{"id":"5_gRbUZa8UZR","colab_type":"code","colab":{}},"source":[""],"execution_count":0,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"5dBFztid8U07","colab_type":"text"},"source":["### 練習問題2. SNP-indexのグラフを描く"]},{"cell_type":"code","metadata":{"id":"K15GXT6X8du1","colab_type":"code","colab":{}},"source":[""],"execution_count":0,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"w9-6gCUyvkSr","colab_type":"text"},"source":["### 練習問題3. SNP-indexのSliding window解析をおこなう\n","イネ第12染色体の全長(bp) = 27,531,856"]},{"cell_type":"code","metadata":{"id":"AN6GWuPOv009","colab_type":"code","colab":{}},"source":[""],"execution_count":0,"outputs":[]}]}