{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "*** IMPORTANT ***\n",
    "Run this cell before practice below.\n",
    "You can download a sample file for this practice.\n",
    "\"\"\"\n",
    "!wget https://raw.githubusercontent.com/CropEvol/lecture/master/data/mutmap_bulk.txt -O mutmap_bulk.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction to Large Data Analysis (1)\n",
    "\n",
    "## Contents\n",
    "\n",
    "### Introduction\n",
    "1. [Genomic analysis and text data](#0.1)\n",
    "1. [Basics of text data processing](#0.2)\n",
    "1. [About sample data](#0.3)\n",
    "\n",
    "--- \n",
    "\n",
    "### Practice\n",
    "1. [Open and close a text-file](#1.1)\n",
    "1. [Read a file line-by-line](#1.2)\n",
    "1. [Remove line feed code](#1.3)\n",
    "1. [Split one line](#1.4)\n",
    "1. [Calculate SNP-index](#1.5)\n",
    "1. [Write into a file](#1.6)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Introduction\n",
    "\n",
    "### Genomic analysis and Text data analyisis<a name=\"0.1\"></a>\n",
    "\n",
    "Big data including genomic data is just text file (but very large data size).  \n",
    "\n",
    "For example, genome sequence data is the strings data consisted of \"A\"(Adenine), \"C\"(Cytosine), \"G\"(Guanine) and \"T\"(Thymine). \n",
    "\n",
    "So, it is not an exaggeration to say that genomic analysis is an analysis that find the impotant data from very long text. \n",
    "\n",
    "In this practice, we learn to deal with large data in a text file by using Python."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Basic of text data processing<a name=\"0.2\"></a>\n",
    "\n",
    "Most important point is ___\"Process lines one by one from the top line\"___\n",
    "\n",
    "Overall flow of text data processing\n",
    "- Open file\n",
    "- Extract one line\n",
    "- Split one line by data\n",
    "- Calculate and/or process the data in one line\n",
    "- Move next line\n",
    "- Close file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sample data => [Link](mutmap_bulk.txt)<a name=\"0.3\"></a>\n",
    "\n",
    "In this practice, we use the MutMap data for the \"pale green\" leaf in rice (Abe et al., 2012). \n",
    "<div style=\"margin-bottom: 5px;\"><img src=\"https://github.com/CropEvol/lecture/blob/master/images/06/mutmap04_en.jpg?raw=true\" alt=\"mutmap\"></div>\n",
    "\n",
    "Abe, A., Kosugi, S., Yoshida, K., Natsume, S., Takagi, H., Kanzaki, H., Matsumura, H., Yoshida, K., Mitsuoka, C., Tamiru, M., Innan, H., Cano, L., Kamoun, S., Terauchi, R. (2012). [Genome sequencing reveals agronomically important loci in rice using MutMap.](https://www.nature.com/articles/nbt.2095) _Nature biotechnology_, 30(2), 174.\n",
    "\n",
    "The MutMap is one of methods of gene mapping.  \n",
    "The outline of MutMap:\n",
    "1. You cross original line (Here is rice cultivar \"Hitomebore\") and its mutate line.\n",
    "1. You develop F2 individuals, and pool the DNA of F2 individuals have mutant phenotype.\n",
    "1. You get two NGS data from original parent line and F2 individuals pooled.\n",
    "1. You compare both data, and you can find position of mutation, what and how many nucleotides are in the position.\n",
    "1. Because it should be that all F2 individuals with mutant phenotype share the same mutation caused the phenotype, you can expect the fixed mutation allele in the locus.\n",
    "1. You search the locus has fixed SNP-index (frequency of mutant allele) in the whole of genome.\n",
    "1. You draw the map of SNP-index (x-axis => Chromsome position, y-axis => SNP-index).\n",
    "1. You can find the candidate locus or site.\n",
    "\n",
    "<div style=\"margin-bottom: 5px;\"><img src=\"https://github.com/CropEvol/lecture/blob/master/images/06/mutmap01_en.jpg?raw=true\" alt=\"mutmap\"></div>\n",
    "\n",
    "<div style=\"margin-bottom: 5px;\"><img src=\"https://github.com/CropEvol/lecture/blob/master/images/06/mutmap02.png?raw=true\" alt=\"mutmap\"></div>\n",
    "\n",
    "Check the sample file => [Link](mutmap_bulk.txt)\n",
    "\n",
    "Information of each columns:\n",
    "- Chromosome\n",
    "- Position\n",
    "- \"Hitomebore\" allele\n",
    "- Mutant allele\n",
    "- Depth of \"Hitomebore\" allele: Counts of \"Hitomebore\" allele in the pooled sequences of 20 individuals have mutant phenotype \n",
    "- Depth of mutant allele: Counts of mutant allele in the pooled sequences of 20 individuals have mutant phenotype \n",
    "\n",
    "<div style=\"margin-bottom: 5px;\"><img src=\"https://github.com/CropEvol/lecture/blob/master/images/06/mutmap03_en.jpg?raw=true\" alt=\"mutmap\"></div>\n",
    "\n",
    "Let's detect the locus has fixed SNP-index (1.0) to find the mutation caused \"pale green\" leaf in rice."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Practice\n",
    "In this practice, we will create the program \"Calculate SNP-index and Write the result into a file\" by adding the code step-by-step.\n",
    "\n",
    "### 1. Open and close a text-file<a name=\"1.1\"></a>\n",
    "Open the sample file. And close the file.\n",
    "\n",
    "=== Basic syntax ===\n",
    "```python\n",
    "f_in = open('<File Name>', 'r')\n",
    "f_in.close()\n",
    "```\n",
    "\n",
    "=== Description ===\n",
    "- open()\n",
    "    - 1st argment: Name of file that you want to open\n",
    "    - 2nd argment: \n",
    "        * 'r': Open a file with read-only mode  \n",
    "        * 'w': Open a file with writing mode\n",
    "        * 'a': Open a file with appending mode\n",
    "    - File contents are set in the variable `f_in`.\n",
    "        (The variable is called __file object__.)\n",
    "- close()\n",
    "    - Usage: FILE_OBJECT.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Add codes to the below ###\n",
    "\n",
    "dataset = 'mutmap_bulk.txt' # File name\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Read a file line-by-line<a name=\"1.2\"></a>\n",
    "Read the opened file line-by-line, and print out strings of the line .\n",
    "\n",
    "=== Basic syntax ===\n",
    "```python\n",
    "for line in f_in:\n",
    "    print(line)\n",
    "```\n",
    "\n",
    "=== Description ===  \n",
    "You can get one line by using `for`.  \n",
    "The contents of `f_in` are like items of list. => [1st line, 2nd line, 3rd line, ..., last line]  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Add codes to the below ###\n",
    "\n",
    "\n",
    "dataset = 'mutmap_bulk.txt' # Input file name\n",
    "\n",
    "f_in = open(dataset, 'r')  # Open the file\n",
    "\n",
    "f_in.close()  # Close the file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Remove line feed code<a name=\"1.3\"></a>\n",
    "\n",
    "In the upper code, you can see the blank lines between each lines.  \n",
    "This is due to each lines in text file include the line feed code(`\\n`).  \n",
    "In text data processing, it is better that the line feed code is removed .\n",
    "\n",
    "=== Basic syntax ===  \n",
    "```python\n",
    "line = line.rstrip()\n",
    "```\n",
    "\n",
    "=== Description ===\n",
    "- `rstrip()`\n",
    "    - Usage: STRINGS_OBJECT.rstrip()\n",
    "    - right-strip: remove a string at the end of strings or an item at the end of list.\n",
    "    - We remove the line feed code at the end of line here.\n",
    "    \n",
    "- `lstrip()`\n",
    "    - Usage: STRINGS_OBJECT.lstrip()\n",
    "    - left-strip: remove the first string of strings or the first item at the first of list.\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Add codes to the below ###\n",
    "\n",
    "dataset = 'mutmap_bulk.txt' #  Input file name\n",
    "\n",
    "f_in   = open(dataset, 'r')  # Open file with read-only mode\n",
    "\n",
    "# Read a file line-by-line\n",
    "for line in f_in:\n",
    "    print(line)\n",
    "    \n",
    "f_in.close()  # Close file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Split one line<a name=\"1.4\"></a>\n",
    "To accesse one data in line, split the line.  \n",
    "This sample file is the tab-delimited file.   \n",
    "You can split the line by specifying the tab code `\\t`.\n",
    "\n",
    "=== Basic syntax ===  \n",
    "```python\n",
    "items = line.split('\\t')\n",
    "```\n",
    "\n",
    "=== Description ===\n",
    "- `split()`\n",
    "    - Usage: STRINGS_OBJECT.split(\"delimiter\")\n",
    "        * Commonly used delimiters\n",
    "            - `\\t`: tab\n",
    "            - `,`: comma\n",
    "    - The split strings is converted to a list.\n",
    "        * ex.) The first line `chr10\t51406\tG\tA\t6\t3` is converted to the list `['chr10', '51406', 'G', 'A', '6', '3']`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Add codes to the below ###\n",
    "\n",
    "dataset = 'mutmap_bulk.txt' #  Input file name\n",
    "\n",
    "f_in   = open(dataset, 'r')  # Open file with read-only mode\n",
    "\n",
    "# Read a file line-by-line\n",
    "for line in f_in:\n",
    "    \n",
    "    line = line.rstrip()  # Remove line feed code\n",
    "    print(line)\n",
    "    \n",
    "f_in.close()  # Close file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5. Calculate SNP-index<a name=\"1.5\"></a>\n",
    "Calculate SNP-index.  \n",
    "You can calculate the following formula.\n",
    "\n",
    "```python\n",
    "SNP-index = Counts of ALT allele / (Counts of REF allele + Counts of ALT allele) \n",
    "\n",
    "\"Counts of REF allele\" is 4th item.\n",
    "\"Counts of ALT allele\" is 5th item.\n",
    "（ATTENTION: \"0-started index\" is used in Python）\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Add codes to the below ###\n",
    "\n",
    "dataset = 'mutmap_bulk.txt' #  Input file name\n",
    "\n",
    "f_in   = open(dataset, 'r')  # Open file with read-only mode\n",
    "\n",
    "# Read a file line-by-line\n",
    "for line in f_in:\n",
    "    \n",
    "    line = line.rstrip()  # Remove line feed code\n",
    "    items = line.split('\\t')  # Split one line\n",
    "    \n",
    "    #print(line)\n",
    "    print(items)\n",
    "    \n",
    "f_in.close()  # Close file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Supplementary: Converting data type\n",
    "\"Counts of REF allele\" and \"Counts of ALT allele\" are extracted as strings from `items`.\n",
    "So, you need to comvert the data type from strings to integer to calculate SNP-index.   \n",
    "\n",
    "=== Basic syntax ===  \n",
    "```python\n",
    "int(Strings or Float number)\n",
    "```\n",
    "\n",
    "=== Description ===\n",
    "- `int()`\n",
    "    - Usage: int(Strings or Float number)\n",
    "    - Converting to integer.\n",
    "- `float()`\n",
    "    -  Converting to float number.\n",
    "- `str()`\n",
    "    - Converting to strings.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6. Write into a file<a name=\"1.6\"></a>\n",
    "\n",
    "Add the calculated SNP-index to the end of line.\n",
    "And write the line to a file.\n",
    "\n",
    "＊ATTENTIONS＊  \n",
    "- To add the SNP-index to strings, you need to convert the value of SNP-index from float number to string. \n",
    "- To print out as tab-delimited line, you need to add tab code `\\t` between the existing line and SNP-index.\n",
    "- Because the line feed code `\\n` is removed at STEP 3, you need to add new `\\n` in the `write()`.\n",
    "\n",
    "=== Basic syntax ===\n",
    "```python\n",
    "f_out = open('<file name>', 'w')\n",
    "f_out.write(STRINGS)　\n",
    "f_out.close()\n",
    "```\n",
    "=== Description ===\n",
    "- write()\n",
    "    - Usage: FILE_OBJECT.write(\"any strings\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Add codes to the below ###\n",
    "\n",
    "dataset = 'mutmap_bulk.txt' #  Input file name\n",
    "\n",
    "f_in   = open(dataset, 'r')  # Open file with read-only mode\n",
    "\n",
    "# Read a file line-by-line\n",
    "for line in f_in:\n",
    "    \n",
    "    line = line.rstrip()  # Remove line feed code\n",
    "    items = line.split('\\t')  # Split one line\n",
    "    snpindex = int(items[5]) / (int(items[4]) + int(items[5]))  # Calculate SNP-index\n",
    "    \n",
    "    #print(line)\n",
    "    #print(items)\n",
    "    #print(items[1])\n",
    "    print(snpindex)\n",
    "    \n",
    "f_in.close()  # Close file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the output file => [File list](./)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Supplementary: Embedding the values of variables into strings printed out\n",
    "\n",
    "If you want to embed the values of variables into strings printed out, you can use the following code.\n",
    "\n",
    "```python\n",
    "x = 5.3\n",
    "y = 3.4\n",
    "\n",
    "print('%d + %d = %d' % (x, y, x + y))\n",
    "# Print out => 5 + 3 = 8\n",
    "\n",
    "print('%f + %.3f = %.2f' % (x, y, x + y))\n",
    "# Print out => 5.300000 + 3.400 = 8.70\n",
    "\n",
    "print('%s + %s = %s' % (x, y, x + y))\n",
    "# Print out => 5.3 + 3.4 = 8.7\n",
    "```\n",
    "\n",
    "In `print()`, there are two parts.\n",
    "ex.) `'%d + %d = %d'`　and `(x, y, x + y)`\n",
    "\n",
    "\n",
    "The former part `'%d + %d = %d'` is a framework of strings printed out.\n",
    "\n",
    "In the framework, there are three `%d`.  \n",
    "Each `%d` in the former part are substituted into `(x, y, x + y)` in the latter part.  \n",
    "- The first `%d` is substituted into `x`.  \n",
    "- The second `%d` is substituted into `y`.  \n",
    "- The third `%d` is substituted into `x + y`.  \n",
    "\n",
    "You can also specify the data type of value to be substituted in by `%d`, `%f` or `%s`.\n",
    "- `%d`: Integer\n",
    "- `%f`: Float number (Usage: %.3f => 0.123; %.8f => 0.12345678)\n",
    "- `%s`: Strings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = 5.3\n",
    "y = 3.4\n",
    "\n",
    "print('%d + %d = %d' % (x, y, x + y))\n",
    "# Print out => 5 + 3 = 8\n",
    "\n",
    "print('%f + %.3f = %.2f' % (x, y, x + y))\n",
    "# Print out => 5.300000 + 3.400 = 8.70\n",
    "\n",
    "print('%s + %s = %s' % (x, y, x + y))\n",
    "# Print out => 5.3 + 3.4 = 8.7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Completed code ###\n",
    "\n",
    "dataset = 'mutmap_bulk.txt' #  Input file name\n",
    "outdata = 'mutmap_snpindex.txt' # Output file name\n",
    "\n",
    "f_in   = open(dataset, 'r')  # Open file with read-only mode\n",
    "f_out = open(outdata, 'w')  # Open file with writing mode\n",
    "\n",
    "# Read a file line-by-line\n",
    "for line in f_in:\n",
    "    \n",
    "    line = line.rstrip()  # Remove line feed code\n",
    "    items = line.split('\\t')  # Split one line\n",
    "    snpindex = int(items[5]) / (int(items[4]) + int(items[5]))  # Calculate SNP-index\n",
    "    \n",
    "    out_line = line + '\\t' + str(snpindex)\n",
    "        # Add SNP-index at the end of line\n",
    "        # Joining Line and SNP-index by tab code because writing out  as tab-delimit text file\n",
    "    \n",
    "    #print(line)\n",
    "    #print(items)\n",
    "    #print(items[1])\n",
    "    #print(snpindex)\n",
    "    print(out_line)\n",
    "    \n",
    "    f_out.write(\"%s\\n\" % out_line)  # Write into the \"f_out\" file\n",
    "    \n",
    "f_in.close()  # Close file\n",
    "f_out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this practice, you write the upper program to calculate SNP-index by using `open()`/`close()`, `for`, `rstrip()`, `split()` and so on.\n",
    "\n",
    "This method can be widely applied. You can use this method to process another large-text data.  \n",
    "But it tend to be longer code using \"Pandas\" library.\n",
    "\n",
    "Using \"Pandas\" library, you can calculate SNP-index easily by less code.\n",
    "\n",
    "In next practice, you will use \"Pandas\" library to calculate SNP-index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
